### ISSUE TITLE
pounce.jax computes Jacobian/Hessian densely despite reporting sparse structure

### ISSUE THREAD

**Issue by jkitchin** [jkitchin/pounce#83]
## Summary

The `pounce.jax` front-end (`from_jax`) reports a *sparse* Jacobian/Hessian structure to the solver but **computes them densely on every evaluation**. The reported sparsity is cosmetic on the storage side only — the actual AD work and memory are `O(m·n)` (Jacobian) and `O(n²)` (Hessian) regardless of how sparse the true matrices are. This issue tracks the design gap and the options to close it.

For contrast, the Rust `.nl` tape path (`crates/pounce-cli/src/nl_tape.rs`, used in `nl_reader.rs:1572-1625`) already does compressed/colored AD for the Hessian. The Python path has none of that.

## Where it happens

`python/pounce/jax/_build.py`:

- Jacobian is built dense via `jax.jacrev` (`_build.py:76`) and, per eval, fully materialized then sliced to the nonzeros (`_build.py:125-127`):

```python
def jacobian(self, x):
    J = _to_np(self._jac_g_dense(jnp.asarray(x)))   # full (m, n) dense
    return J[self._jac_rows, self._jac_cols]          # discard everything else
```

- Hessian of the Lagrangian is the same story — dense `(n, n)` then sliced to the lower triangle (`_build.py:132-137`).

- `jax.jacrev`/`jacfwd`/`hessian` never produce sparse output; they are dense by construction. So sparsity has to be *computed*, not requested.

### Cost example
A 10,000×10,000 constraint system with ~5 nonzeros/row has ~50k true Jacobian entries but currently computes 10^8 every iteration.

## Secondary issue: pattern detection is a single random probe

Sparsity is detected from one random probe (`_build.py:96-109`), threshold `1e-12`. With the current dense-slice approach a mis-probe only drops a *reported* nonzero. If we move to compressed/colored AD (below), the pattern is baked into the seed structure, so a wrong probe also corrupts the compression/decoding. Value-dependent sparsity (branchy `where`/`abs` structure) is the risk case. Robustifying detection (union of several probes, or symbolic) becomes more important once compression is in play.

## Options to close the gap (escalating)

**A. Mode selection (cheap, partial).** `jacrev` costs ~`m` passes, `jacfwd` ~`n`. Pick `jacfwd` when `n < m`. Still dense, just stops over-paying on tall/skinny shapes.

**B. Compressed (colored) AD — the real fix.** CPR-style: color structurally-orthogonal columns, do one JVP per color (`k ≪ n` colors), `vmap` over seeds → `(m, k)` compressed matrix, scatter back to the known nonzeros. Cost drops from `O(n)` to `O(k)` passes. This mirrors what the Rust Hessian path already does, so it keeps both front-ends architecturally consistent. The pattern pounce already detects (`self._jac_rows/_jac_cols`) is exactly the input coloring needs.

**C. `sparsejac` (least code).** The `sparsejac` library implements B on top of `jax.experimental.sparse.BCOO` + graph coloring. Hand it the existing pattern as a `BCOO` template; it returns a function producing a sparse Jacobian. Cost: a third-party dep + reconciling `BCOO.data` ordering with pounce's `(rows, cols)` ordering.

## Suggested path

1. Short term: Option A (one-liner) to stop pathological over-computation.
2. Medium term: Option B for the Jacobian and Hessian, reusing the coloring approach from the Rust tape path; gate behind a flag and benchmark dense-vs-sparse on an existing test problem.
3. Alongside B: harden pattern detection (multi-probe union) since compression makes a mis-probe more costly.

Environment when filed: jax 0.10.0, `jax.experimental.sparse` (BCOO/BCSR) available.
